backend/remote-state/oss/client.go

```diff
@@ -0,0 +1,232 @@
+package oss
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/aliyun/aliyun-oss-go-sdk/oss"
+	multierror "github.com/hashicorp/go-multierror"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+	"log"
+)
+
+type RemoteClient struct {
+	ossClient            *oss.Client
+	bucketName           string
+	statePath            string
+	lockPath             string
+	serverSideEncryption bool
+	acl                  string
+}
+
+func (c *RemoteClient) Get() (payload *remote.Payload, err error) {
+	buf, err := c.getObj(c.statePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there was no data, then return nil
+	if buf == nil || len(buf.Bytes()) == 0 {
+		log.Printf("State %s has no data.", c.statePath)
+		return nil, nil
+	}
+
+	var hashChannel = make(chan []byte, 1)
+	sum := md5.Sum(buf.Bytes())
+	hashChannel <- sum[:]
+	payload = &remote.Payload{
+		Data: buf.Bytes(),
+		MD5:  <-hashChannel,
+	}
+	return payload, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	return c.putObj(c.statePath, data)
+}
+
+func (c *RemoteClient) Delete() error {
+	return c.deleteObj(c.statePath)
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	bucket, err := c.ossClient.Bucket(c.bucketName)
+	if err != nil {
+		return "", fmt.Errorf("Error getting bucket: %#v", err)
+	}
+
+	log.Printf("Lock info:%#v", info)
+
+	infoJson, err := json.Marshal(info)
+	if err != nil {
+		return "", err
+	}
+
+	if info.ID == "" {
+		lockID, err := uuid.GenerateUUID()
+		if err != nil {
+			return "", err
+		}
+		info.ID = lockID
+	}
+	info.Path = c.lockPath
+	if exist, err := bucket.IsObjectExist(info.Path); err != nil {
+		return "", fmt.Errorf("Estimating object %s is exist got an error: %#v", info.Path, err)
+	} else if !exist {
+		if err := c.putObj(info.Path, infoJson); err != nil {
```

Comment:
Checking the existence then putting the lock object creates a race condition. 
Is there some way to write the object and have it rejected the request if it exists? (this is one of the reasons the s3 backend has to delegate locking to DynamoDB)

Reply:
This still applies